src/back/hlsl/writer.rs

```diff
@@ -1023,18 +1032,74 @@ impl<'a, W: Write> Writer<'a, W> {
             }
             Expression::ImageSample {
                 image,
-                sampler,        // TODO:
-                coordinate,     // TODO:
-                array_index: _, // TODO:
-                offset: _,      // TODO:
-                level: _,       // TODO:
-                depth_ref: _,   // TODO:
+                sampler,
+                coordinate,
+                array_index,
+                offset,
+                level,
+                depth_ref,
             } => {
+                use crate::SampleLevel as Sl;
+
+                let texture_func = match level {
+                    Sl::Auto => {
+                        if depth_ref.is_some() {
+                            "SampleCmp"
+                        } else {
+                            "Sample"
+                        }
+                    }
+                    Sl::Zero => "SampleCmpLevelZero",
+                    Sl::Exact(_) => "SampleLevel",
+                    Sl::Bias(_) => "SampleBias",
+                    Sl::Gradient { .. } => "SampleGrad",
+                };
+
                 self.write_expr(module, image, func_ctx)?;
-                write!(self.out, ".Sample(")?;
+                write!(self.out, ".{}(", texture_func)?;
                 self.write_expr(module, sampler, func_ctx)?;
                 write!(self.out, ", ")?;
                 self.write_expr(module, coordinate, func_ctx)?;
+
+                if let Some(array_index) = array_index {
+                    write!(self.out, ", ")?;
+                    self.write_expr(module, array_index, func_ctx)?;
+                }
+
+                if let Some(depth_ref) = depth_ref {
+                    write!(self.out, ", ")?;
+                    self.write_expr(module, depth_ref, func_ctx)?;
+                }
+
+                match level {
+                    Sl::Auto => {}
+                    Sl::Zero => {
+                        // Level 0 is implied for depth comparison
```

Comment:
since `SampleCmpLevelZero` is used, why do we need to add `0.0` even for non-depth?